File: X_Ray_Rust_Full/src/similarity.rs
// src/similarity.rs — Similarity computation engine (unified, zero duplication)
//
// Provides: code_similarity, name_similarity, signature_similarity,
//           callgraph_overlap, semantic_similarity
//
// Uses n-gram fingerprinting (winnowed, MOSS-style) + token histogram.

use crate::tokenizer;
use sha2::{Digest, Sha256};
use std::collections::{HashMap, HashSet};
use regex::Regex;
use std::sync::LazyLock;

static RE_PY_TOKEN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?x)
        \b(def|class|if|elif|else|for|while|try|except|finally|with|
           return|yield|import|from|as|in|not|and|or|is|lambda|
           raise|pass|break|continue|assert|global|nonlocal|async|await)\b |
        \b(True|False|None|print|len|range|int|str|float|list|dict|set|tuple|
           type|isinstance|hasattr|getattr|setattr|enumerate|zip|map|filter|
           sorted|reversed|any|all|min|max|sum|abs|round|open|super)\b |
        \b[A-Za-z_]\w*\b |
        \b\d+(?:\.\d+)?\b |
        "[^"]*"|'[^']*' |
        [+\-*/=%<>!&|^~@]+|[(){}\[\],;:.]
    "#).unwrap()
});

// Token categories for normalization
fn classify_token(tok: &str) -> Option<String> {
    // Keywords: keep as-is
    let keywords: HashSet<&str> = [
        "def","class","if","elif","else","for","while","try","except",
        "finally","with","return","yield","import","from","as","in","not",
        "and","or","is","lambda","raise","pass","break","continue","assert",
        "global","nonlocal","async","await",
    ].iter().copied().collect();

    let builtins: HashSet<&str> = [
        "True","False","None","print","len","range","int","str","float",
        "list","dict","set","tuple","type","isinstance","hasattr","getattr",
        "setattr","enumerate","zip","map","filter","sorted","reversed","any",
        "all","min","max","sum","abs","round","open","super",
    ].iter().copied().collect();

    if keywords.contains(tok) {
        Some(tok.to_string())
    } else if builtins.contains(tok) {
        Some(tok.to_string())
    } else if tok.chars().next().map_or(false, |c| c.is_alphabetic() || c == '_') {
        Some("ID".to_string())
    } else if tok.chars().next().map_or(false, |c| c.is_ascii_digit()) {
        Some("NUM".to_string())
    } else if tok.starts_with('"') || tok.starts_with('\'') {
        Some("STR".to_string())
    } else {
        Some(tok.to_string()) // operators, punctuation
    }
}

/// Produce normalized token stream from Python source
fn normalized_token_stream(code: &str) -> Vec<String> {
    RE_PY_TOKEN
        .find_iter(code)
        .filter_map(|m| classify_token(m.as_str()))
        .collect()
}

/// Winnowed n-gram fingerprints (MOSS algorithm)
fn ngram_fingerprints(tokens: &[String], n: usize, w: usize) -> HashSet<u64> {
    if tokens.len() < n {
        return HashSet::new();
    }
    // Hash each n-gram
    let hashes: Vec<u64> = (0..=tokens.len() - n)
        .map(|i| {
            let gram = tokens[i..i + n].join(" ");
            let mut hasher = Sha256::new();
            hasher.update(gram.as_bytes());
            let result = hasher.finalize();
            u64::from_be_bytes(result[..8].try_into().unwrap())
        })
        .collect();

    if hashes.len() < w {
        return hashes.into_iter().collect();
    }

    // Winnowing: keep minimum from each window
    let mut fingerprints = HashSet::new();
    for window in hashes.windows(w) {
        if let Some(&min_hash) = window.iter().min() {
            fingerprints.insert(min_hash);
        }
    }
    fingerprints
}

/// Token n-gram similarity (Jaccard of winnowed fingerprints)
fn token_ngram_similarity(code_a: &str, code_b: &str) -> f64 {
    let toks_a = normalized_token_stream(code_a);
    let toks_b = normalized_token_stream(code_b);
    let fp_a = ngram_fingerprints(&toks_a, 5, 4);
    let fp_b = ngram_fingerprints(&toks_b, 5, 4);

    if fp_a.is_empty() && fp_b.is_empty() {
        return 0.0;
    }
    let inter = fp_a.intersection(&fp_b).count() as f64;
    let union = fp_a.union(&fp_b).count() as f64;
    if union == 0.0 { 0.0 } else { inter / union }
}

/// AST node-type histogram similarity
fn ast_histogram(code: &str) -> HashMap<String, u32> {
    let re = Regex::new(r"(?m)^\s*(def|class|if|elif|else|for|while|try|except|finally|with|return|yield|import|from|raise|pass|break|continue|assert|async|await)\b").unwrap();
    let mut counts = HashMap::new();
    for cap in re.captures_iter(code) {
        let kw = cap.get(1).unwrap().as_str();
        *counts.entry(kw.to_string()).or_insert(0) += 1;
    }
    // Also count assignments, calls, and operators
    let re_assign = Regex::new(r"(?m)^\s*\w+\s*=").unwrap();
    *counts.entry("assign".to_string()).or_insert(0) += re_assign.find_iter(code).count() as u32;
    counts
}

fn ast_histogram_similarity(code_a: &str, code_b: &str) -> f64 {
    let hist_a = ast_histogram(code_a);
    let hist_b = ast_histogram(code_b);
    tokenizer::cosine_similarity(&hist_a, &hist_b)
}

/// Combined code similarity (weighted: 35% n-gram + 65% AST histogram)
pub fn code_similarity(code_a: &str, code_b: &str) -> f64 {
    let ngram = token_ngram_similarity(code_a, code_b);
    let ast = ast_histogram_similarity(code_a, code_b);
    (0.35 * ngram + 0.65 * ast).min(1.0)
}

/// Name similarity (Jaccard of tokenized name parts)
pub fn name_similarity(name_a: &str, name_b: &str) -> f64 {
    let toks_a: HashSet<String> = tokenizer::tokenize(name_a).into_iter().collect();
    let toks_b: HashSet<String> = tokenizer::tokenize(name_b).into_iter().collect();
    tokenizer::jaccard(&toks_a, &toks_b)
}

/// Signature similarity (params + return type + async match)
pub fn signature_similarity(
    params_a: &[String], params_b: &[String],
    ret_a: &Option<String>, ret_b: &Option<String>,
    async_a: bool, async_b: bool,
) -> f64 {
    // Jaccard of param name tokens
    let pa: HashSet<String> = params_a.iter().flat_map(|p| tokenizer::tokenize(p)).collect();
    let pb: HashSet<String> = params_b.iter().flat_map(|p| tokenizer::tokenize(p)).collect();
    let param_jaccard = tokenizer::jaccard(&pa, &pb);

    // Param count ratio
    let max_params = params_a.len().max(params_b.len());
    let param_ratio = if max_params == 0 {
        1.0
    } else {
        params_a.len().min(params_b.len()) as f64 / max_params as f64
    };

    // Return type match
    let ret_match = match (ret_a, ret_b) {
        (Some(a), Some(b)) if a == b => 1.0,
        (None, None) => 0.5,
        _ => 0.0,
    };

    // Async match
    let async_match = if async_a == async_b { 1.0 } else { 0.0 };

    (param_jaccard + param_ratio + ret_match + async_match) / 4.0
}

/// Call-graph overlap (Jaccard of calls_to sets)
pub fn callgraph_overlap(calls_a: &[String], calls_b: &[String]) -> f64 {
    let sa: HashSet<String> = calls_a.iter().cloned().collect();
    let sb: HashSet<String> = calls_b.iter().cloned().collect();
    tokenizer::jaccard(&sa, &sb)
}

/// Full semantic similarity (weighted composite)
pub fn semantic_similarity(
    f1: &crate::types::FunctionRecord,
    f2: &crate::types::FunctionRecord,
) -> f64 {
    let name_sim = name_similarity(&f1.name, &f2.name);
    let sig_sim = signature_similarity(
        &f1.parameters, &f2.parameters,
        &f1.return_type, &f2.return_type,
        f1.is_async, f2.is_async,
    );
    let call_sim = callgraph_overlap(&f1.calls_to, &f2.calls_to);

    // Docstring similarity
    let doc_sim = match (&f1.docstring, &f2.docstring) {
        (Some(d1), Some(d2)) => {
            let tf1 = tokenizer::term_freq(&tokenizer::tokenize(d1));
            let tf2 = tokenizer::term_freq(&tokenizer::tokenize(d2));
            tokenizer::cosine_similarity(&tf1, &tf2)
        }
        _ => 0.0,
    };

    0.30 * name_sim + 0.25 * sig_sim + 0.30 * call_sim + 0.15 * doc_sim
}

/// Quick token cosine for pre-filtering
pub fn quick_token_cosine(code_a: &str, code_b: &str) -> f64 {
    let tf_a = tokenizer::term_freq(&tokenizer::tokenize(code_a));
    let tf_b = tokenizer::term_freq(&tokenizer::tokenize(code_b));
    tokenizer::cosine_similarity(&tf_a, &tf_b)
}

File: Analysis/similarity.py
import re
import hashlib
import keyword
import io
import tokenize as _tokenize_mod
import ast
from collections import Counter
from typing import List

# Optional Rust acceleration (preserved from original)
try:
    import x_ray_core as _rust_core

    _HAS_RUST = True
except ImportError:
    _rust_core = None
    _HAS_RUST = False

# Import canonical definitions — single source of truth
from Core.config import _BUILTIN_NAMES
from Lang.tokenizer import tokenize, term_freq as _term_freq, cosine_similarity

_SPLIT_RE = re.compile(r"[A-Z]+(?=[A-Z][a-z])|[A-Z]?[a-z]+|[A-Z]+|[0-9]+")


# tokenize, _term_freq, cosine_similarity are imported from Lang.tokenizer


def _classify_name(name: str) -> str:
    """Classify a NAME token as keyword, builtin, or generic identifier."""
    if keyword.iskeyword(name) or keyword.issoftkeyword(name):
        return name
    if name in _BUILTIN_NAMES:
        return name
    return "ID"


_TOKEN_TYPE_MAP = {
    _tokenize_mod.NUMBER: "NUM",
    _tokenize_mod.STRING: "STR",
}

_PASSTHROUGH_TYPES = frozenset(
    {
        _tokenize_mod.OP,
        _tokenize_mod.NEWLINE,
        _tokenize_mod.INDENT,
        _tokenize_mod.DEDENT,
    }
)


def _classify_token(tok) -> str | None:
    """Map a single token to its normalized form, or None to skip."""
    if tok.type == _tokenize_mod.NAME:
        return _classify_name(tok.string)
    if tok.type in _TOKEN_TYPE_MAP:
        return _TOKEN_TYPE_MAP[tok.type]
    if tok.type in _PASSTHROUGH_TYPES:
        return tok.string
    return None


def _normalized_token_stream(code: str) -> List[str]:
    """Tokenize Python code, normalizing identifiers and literals."""
    tokens: List[str] = []
    try:
        for tok in _tokenize_mod.generate_tokens(io.StringIO(code).readline):
            normalized = _classify_token(tok)
            if normalized is not None:
                tokens.append(normalized)
    except _tokenize_mod.TokenError:
        pass
    return tokens


def _ngram_fingerprints(tokens: List[str], n: int = 5, w: int = 4) -> frozenset:
    """Winnowed n-gram fingerprinting (MOSS algorithm)."""
    if len(tokens) < n:
        return frozenset()
    hashes: List[int] = []
    for i in range(len(tokens) - n + 1):
        gram = " ".join(tokens[i : i + n])
        hashes.append(int(hashlib.sha256(gram.encode()).hexdigest()[:8], 16))
    if len(hashes) < w:
        return frozenset(hashes)
    fingerprints: set = set()
    for i in range(len(hashes) - w + 1):
        fingerprints.add(min(hashes[i : i + w]))
    return frozenset(fingerprints)


def _token_ngram_similarity(code_a: str, code_b: str) -> float:
    """Jaccard similarity of winnowed n-gram fingerprints."""
    fp_a = _ngram_fingerprints(_normalized_token_stream(code_a))
    fp_b = _ngram_fingerprints(_normalized_token_stream(code_b))
    if not fp_a or not fp_b:
        return 0.0
    return len(fp_a & fp_b) / len(fp_a | fp_b)


def _ast_node_histogram(code: str) -> Counter:
    """Count occurrences of each AST node type."""
    try:
        tree = ast.parse(code)
    except SyntaxError:
        return Counter()
    return Counter(type(node).__name__ for node in ast.walk(tree))


# cosine_similarity is imported from Lang.tokenizer (clamped version)


def _ast_histogram_similarity(code_a: str, code_b: str) -> float:
    """Cosine similarity of AST node-type histograms."""
    ha = _ast_node_histogram(code_a)
    hb = _ast_node_histogram(code_b)
    return cosine_similarity(ha, hb)


def code_similarity(code_a: str, code_b: str) -> float:
    """
    Structural similarity between two code blocks (0–1).
    Uses Python implementation by default, falls back to Rust if configured.
    """
    if _HAS_RUST:
        return _rust_core.code_similarity(code_a, code_b)
    if not code_a or not code_b:
        return 0.0
    if code_a == code_b:
        return 1.0
    tok_sim = _token_ngram_similarity(code_a, code_b)
    ast_sim = _ast_histogram_similarity(code_a, code_b)
    return min(0.35 * tok_sim + 0.65 * ast_sim, 1.0)


def name_similarity(name_a: str, name_b: str) -> float:
    """Semantic similarity between two function names (0–1)."""
    ta = set(tokenize(name_a))
    tb = set(tokenize(name_b))
    if not ta or not tb:
        return 0.0
    return len(ta & tb) / len(ta | tb)


def signature_similarity(func_a, func_b) -> float:
    """Compare two function signatures.
    Note: func_a/func_b are FunctionRecord objects (passed as duck types here to avoid circular dependencies).
    """
    scores: List[float] = []

    # Param-name overlap
    pa = set(tokenize(" ".join(func_a.parameters)))
    pb = set(tokenize(" ".join(func_b.parameters)))
    if pa or pb:
        scores.append(len(pa & pb) / len(pa | pb) if (pa | pb) else 0.0)
    else:
        scores.append(1.0)

    # Param-count ratio
    la, lb = len(func_a.parameters), len(func_b.parameters)
    if max(la, lb) > 0:
        scores.append(min(la, lb) / max(la, lb))
    else:
        scores.append(1.0)

    # Return type match
    ra = (func_a.return_type or "").lower()
    rb = (func_b.return_type or "").lower()
    if ra and rb:
        scores.append(1.0 if ra == rb else 0.0)
    elif not ra and not rb:
        scores.append(0.5)
    else:
        scores.append(0.0)

    # Async-flag match
    scores.append(1.0 if func_a.is_async == func_b.is_async else 0.0)

    return sum(scores) / len(scores) if scores else 0.0


def callgraph_overlap(func_a, func_b) -> float:
    """Jaccard overlap of functions each function calls."""
    ca = set(func_a.calls_to)
    cb = set(func_b.calls_to)
    if not ca or not cb:
        return 0.0
    return len(ca & cb) / len(ca | cb)


def semantic_similarity(func_a, func_b) -> float:
    """Weighted composite of behavioural signals (0–1)."""
    w_name = 0.30
    w_sig = 0.25
    w_call = 0.30
    w_doc = 0.15

    ns = name_similarity(func_a.name, func_b.name)
    ss = signature_similarity(func_a, func_b)
    cg = callgraph_overlap(func_a, func_b)

    # Docstring similarity (token cosine)
    da = _term_freq(tokenize(func_a.docstring or ""))
    db = _term_freq(tokenize(func_b.docstring or ""))
    ds = cosine_similarity(da, db) if (da and db) else 0.0

    return w_name * ns + w_sig * ss + w_call * cg + w_doc * ds
